C error occurs on next()

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: linux, Ubuntu 20.04.3 LTS (WSL 2)
Processor: 4 cpus, 64bit, little endian, AMD Ryzen 3 2200G with Radeon Vega Graphics
CC version: cc (Ubuntu 11.1.0-1ubuntu1~20.04) 11.1.0

getwd: /home/simon/osu/ragnarok.v
vmodules: /home/simon/.vmodules
vroot: /home/simon/v
vexe: /home/simon/v/v
vexe mtime: 2021-11-20 17:37:12
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 3caeadf.82010e7

Git version: git version 2.25.1
Git vroot status: weekly.2021.46-55-g82010e72-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de
```

**What did you do?**
```v
pub fn (mut r Reader) next() ?Packet {
	for {
		if r.pos + 7 > r.buffer().len {
			break
		}
		
		r.pid = r.read_i16()
		r.pos += 1
		r.plen = r.read_i32()

		if r.pid !in glob_packets {
			log.warn("Unhandled packet id: $r.pid")

			if r.plen != 0 {
				r.pos += r.plen
			}
		} else {
			break
		}
	}

	return glob_packets[r.pid]
}
```
```v
for mut packet in r {
        packet.handle(mut p, mut r)
}
```



**What did you expect to see?**
No C error
 
**What did you see instead?**
```
/tmp/v_1000/ragnarok.8769377413119392899.tmp.c:41429: error: struct or union expected
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

---
I can print out `packet`, I just can't access it's fields.
```v
packet.io.Packet{
    packet: osu_change_action
    handle: fn (mut objects.Playermut , packet.io.Reader)
    restricted: true
}
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3395,7 +3395,14 @@ pub fn (mut c Checker) selector_expr(mut node ast.SelectorExpr) ast.Type {
 
 	old_selector_expr := c.inside_selector_expr
 	c.inside_selector_expr = true
-	typ := c.expr(node.expr)
+	mut typ := c.expr(node.expr)
+	if node.expr.is_auto_deref_var() {
+		if mut node.expr is ast.Ident {
+			if mut node.expr.obj is ast.Var {
+				typ = node.expr.obj.typ
+			}
+		}
+	}
 	c.inside_selector_expr = old_selector_expr
 	c.using_new_err_struct = using_new_err_struct_save
 	if typ == ast.void_type_idx {
@@ -4857,7 +4864,10 @@ fn (mut c Checker) for_in_stmt(mut node ast.ForInStmt) {
 			if next_fn.params.len != 1 {
 				c.error('iterator method `next()` must have 0 parameters', node.cond.position())
 			}
-			val_type := next_fn.return_type.clear_flag(.optional)
+			mut val_type := next_fn.return_type.clear_flag(.optional)
+			if node.val_is_mut {
+				val_type = val_type.ref()
+			}
 			node.cond_type = typ
 			node.kind = sym.kind
 			node.val_type = val_type

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2183,7 +2183,7 @@ fn (mut g Gen) for_in_stmt(node ast.ForInStmt) {
 		val := if node.val_var in ['', '_'] { g.new_tmp_var() } else { node.val_var }
 		val_styp := g.typ(node.val_type)
 		if node.val_is_mut {
-			g.writeln('\t$val_styp* $val = ($val_styp*)${t_var}.data;')
+			g.writeln('\t$val_styp $val = ($val_styp)${t_var}.data;')
 		} else {
 			g.writeln('\t$val_styp $val = *($val_styp*)${t_var}.data;')
 		}

--- a/vlib/v/tests/for_in_mut_iterator_val_test.v
+++ b/vlib/v/tests/for_in_mut_iterator_val_test.v
@@ -1,0 +1,50 @@
+const (
+	packets = {
+		0: &Packet{
+			pid: 2
+		}
+		1: &Packet{
+			pid: 1
+		}
+	}
+)
+
+struct Packet {
+	pid        int
+	handle     fn () string
+	restricted bool
+}
+
+struct Reader {
+	foo int
+	bar string
+mut:
+	index int
+}
+
+fn (mut p Reader) next() ?&Packet {
+	if p.index + 1 > packets.len {
+		return error('')
+	}
+	if p.index !in packets {
+		return error('')
+	}
+
+	p.index++
+	return packets[p.index - 1]
+}
+
+fn test_for_in_mut_interator_val() {
+	r := Reader{}
+	mut rets := []string{}
+
+	for mut packet in r {
+		println(packet.pid)
+		rets << '$packet.pid'
+	}
+
+	println(rets)
+	assert rets.len == 2
+	assert rets[0] == '2'
+	assert rets[1] == '1'
+}

